contracts/math/andromeda-time-gate/src/contract.rs

```diff
@@ -0,0 +1,295 @@
+
+const DEFAULT_TIME_INTERVAL: u64 = 3600;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let resp = ADOContract::default().instantiate(
+        deps.storage,
+        env.clone(),
+        deps.api,
+        &deps.querier,
+        info,
+        BaseInstantiateMsg {
+            ado_type: CONTRACT_NAME.to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            kernel_address: msg.kernel_address,
+            owner: msg.owner,
+        },
+    )?;
+
+    let cycle_start_time_milliseconds = match msg.cycle_start_time.clone() {
+        None => Milliseconds::from_nanos(env.block.time.nanos()),
+        Some(start_time) => start_time.get_time(&env.block),
+    };
+
+    let (cycle_start_time, _) = get_and_validate_start_time(&env, msg.cycle_start_time)?;
+
+    let time_interval_seconds = msg.time_interval.unwrap_or(DEFAULT_TIME_INTERVAL);
+
+    GATE_ADDRESSES.save(deps.storage, &msg.gate_addresses)?;
+    CYCLE_START_TIME.save(
+        deps.storage,
+        &(cycle_start_time, cycle_start_time_milliseconds),
+    )?;
+    TIME_INTERVAL.save(deps.storage, &time_interval_seconds)?;
+
+    Ok(resp)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let ctx = ExecuteContext::new(deps, info, env);
+    match msg {
+        ExecuteMsg::AMPReceive(pkt) => {
+            ADOContract::default().execute_amp_receive(ctx, pkt, handle_execute)
+        }
+        _ => handle_execute(ctx, msg),
+    }
+}
+
+fn handle_execute(mut ctx: ExecuteContext, msg: ExecuteMsg) -> Result<Response, ContractError> {
+    let action = msg.as_ref().to_string();
+
+    let action_response = call_action(
+        &mut ctx.deps,
+        &ctx.info,
+        &ctx.env,
+        &ctx.amp_ctx,
+        msg.as_ref(),
+    )?;
+
+    let res = match msg {
+        ExecuteMsg::UpdateCycleStartTime { cycle_start_time } => {
+            execute_update_cycle_start_time(ctx, cycle_start_time, action)
+        }
+        ExecuteMsg::UpdateGateAddresses { new_gate_addresses } => {
+            execute_update_gate_addresses(ctx, new_gate_addresses, action)
+        }
+        ExecuteMsg::UpdateTimeInterval { time_interval } => {
+            execute_update_time_interval(ctx, time_interval, action)
+        }
+        _ => ADOContract::default().execute(ctx, msg),
+    }?;
+
+    Ok(res
+        .add_submessages(action_response.messages)
+        .add_attributes(action_response.attributes)
+        .add_events(action_response.events))
+}
+
+fn execute_update_cycle_start_time(
+    ctx: ExecuteContext,
+    cycle_start_time: Option<Expiry>,
+    action: String,
+) -> Result<Response, ContractError> {
+    let ExecuteContext { deps, info, .. } = ctx;
+
+    ensure!(
```

Comment:
After merging with `main` replace this with the attribute `restricted` (on the ExecuteMsg lvl). The same way it's done other contracts